src/server/dflycmd.cc

```diff
@@ -614,27 +628,29 @@ void DflyCmd::StopReplication(uint32_t sync_id) {
 }
 
 void DflyCmd::CancelReplication(uint32_t sync_id, shared_ptr<ReplicaInfo> replica_ptr) {
-  lock_guard lk(replica_ptr->mu);
-  if (replica_ptr->replica_state == SyncState::CANCELLED) {
-    return;
-  }
+  {
+    lock_guard lk = replica_ptr->GetExclusiveLock();
+    if (replica_ptr->replica_state == SyncState::CANCELLED) {
+      return;
+    }
 
```

Comment:
ok, so the main change is that we release the mutex now, makes sense as it's not needed for any of the steps below